api/admin/controller.py

```diff
@@ -3316,228 +3335,101 @@ def discovery_services(self):
         if flask.request.method == "GET":
+            # Make a list of all discovery services, each with the
+            # list of libraries registered with that service and the
+            # status of the registration.
             services = []
-            for service in self._db.query(ExternalIntegration).filter(
-                ExternalIntegration.goal==ExternalIntegration.DISCOVERY_GOAL):
-
+            for registry in RemoteRegistry.for_protocol_and_goal(
+                    self._db, ExternalIntegration.OPDS_REGISTRATION, goal
+            ):
                 libraries = []
-                for library in service.libraries:
+                for registration in registry.registrations:
+                    library = registration.library
                     library_info = dict(short_name=library.short_name)
-                    status = ConfigurationSetting.for_library_and_externalintegration(
-                        self._db, LIBRARY_REGISTRATION_STATUS, library, service).value
+                    status = registration.status_field.value
                     if status:
                         library_info["status"] = status
                         libraries.append(library_info)
 
                 services.append(
                     dict(
-                        id=service.id,
+                        id=registry.integration.id,
                         libraries=libraries,
                     )
                 )
-
             return dict(library_registrations=services)
 
         if flask.request.method == "POST":
-
+            # Attempt to register a library with a RemoteRegistry.
             integration_id = flask.request.form.get("integration_id")
             library_short_name = flask.request.form.get("library_short_name")
-            stage = flask.request.form.get("registration_stage")
+            stage = flask.request.form.get("registration_stage") or Registration.TESTING_STAGE
 
-            integration = get_one(self._db, ExternalIntegration,
-                                  goal=ExternalIntegration.DISCOVERY_GOAL,
-                                  id=integration_id)
-            if not integration:
+            registry = RemoteRegistry.for_integration_id(
+                self._db, integration_id, goal
+            )
+            if not registry:
                 return MISSING_SERVICE
 
             library = get_one(self._db, Library, short_name=library_short_name)
             if not library:
                 return NO_SUCH_LIBRARY
 
-            integration.libraries += [library]
-            status = ConfigurationSetting.for_library_and_externalintegration(
-                self._db, LIBRARY_REGISTRATION_STATUS, library, integration)
-            status.value = FAILURE
-            registered = self._register_library(integration.url, library, integration, stage=stage, do_get=do_get, do_post=do_post, key=key)
+            registration = Registration(registry, library)
```

Comment:
In the old code, we set the registration status back to failure when you attempted to re-register. That probably wasn't great, since you might still be registered even if the update failed. But we should have some way to track that the update failed.

Reply:
I changed the code to do this consistently inside registry.py -- it was working the way you said for shared collections but not for the library registry.

I agree that later on we should figure out a way to distinguish between "reregistration failed but you're still good" and "this went so badly that you're no longer registered".